### ISSUE TITLE
Power Spectra Not Showing for MELODIC Overlays

### ISSUE THREAD

**Issue by Nagle2036** [pauldmccarthy/fsleyes#106]
Hi there,

I see that on 16th July 2021, there was a fix to enable the x-axis of the power spectrum view to show frequencies of ICA power spectra for MELODIC overlays. 

However, when loading my melodic_IC.nii.gz file into Fsleyes in the Melodic scene view, the power spectrum does not show (please see the screenshot attached, forgive the phone pic). A spectrum appears if I untick either 'Show frequencies along x axis' or 'Plot component power spectra for Melodic images' in the power spectrum control settings window, but I don't think this is showing me what I need for my purposes of IC classification into noise vs signal. 

Would appreciate any guidance on this, thank you!

Alex

![IMG_20230310_112158](https://user-images.githubusercontent.com/117359984/224304659-a2c09a86-45f8-4af5-b728-8597db67c324.jpg)

**Comment by pauldmccarthy** [comment]
Hi @Nagle2036, can I ask what version of FSLeyes you are using? There was a change in 1.5.0 related to this behaviour, so if you are using an older version I would suggest updating to the latest available.

**Comment by Nagle2036** [comment]
Thanks Paul for the quick reply. Running `fsleyes --version` in my command line says I have version 1.5.0, so the change should be accounted for right? Or would you still suggest updating to a more recent version?

Thanks,
Alex

**Comment by pauldmccarthy** [comment]
Hmm, there is a chance that the bug is still present in some form. Would you be able to do a couple of things for me:

1. Could you open _Settings -> Power spectra -> Overlay List_, and check whether the `melodic_IC` overlay is hidden (via the eye button)?

2. Can you describe how you opened FSLeyes and melodic mode, and loaded the images - e.g. did you open FSLeyes, then load the images, then open melodic mode? Or did you do things differently or in a different order?

**Comment by Nagle2036** [comment]
Sure. 

1. The melodic_IC overlay is definitely enabled in the Overlay List, as shown with the eye button.

2. I opened FSLeyes and the files simultaneously from the command line using `fsleyes --scene melodic -ad p427.ica/filtered_func_data.ica`

Hope that's helpful. Thank you.

Alex

**Comment by pauldmccarthy** [comment]
@Nagle2036 Hopefully I'll have a new release out within the next month or so, but that is very much contingent on how many interruptions occur. In the meantime, you could patch your own installation, as the fix is quite easy.  Open `<your-python-env>/lib/python<X.Y>/site-packages/fsleyes/plotting/powerspectrumseries.py` in a text editor, and make this change:

```diff

diff --git a/fsleyes/plotting/powerspectrumseries.py b/fsleyes/plotting/powerspectrumseries.py
index 240d4925..34a52ff1 100644
--- a/fsleyes/plotting/powerspectrumseries.py
+++ b/fsleyes/plotting/powerspectrumseries.py
@@ -485,9 +485,17 @@ class MelodicPowerSpectrumSeries(dataseries.DataSeries,
         opts      = self.displayCtx.getOpts(overlay)
         component = opts.volume

-        ydata     = overlay.getComponentPowerSpectrum(component)
-        nsamples  = len(overlay.getComponentTimeSeries(component))
-        xdata     = calcFrequencies(nsamples, overlay.tr, np.float32)
+        ydata    = overlay.getComponentPowerSpectrum(component)
+        nsamples = len(overlay.getComponentTimeSeries(component))
+
+        # Melodic always outputs an even number of frequency
+        # values, regardless of whether the input was odd or
+        # even. calcFrequencies produces floor(nsamples / 2)
+        # values, so we add one to odd nsamples to compensate.
+        if nsamples % 2:
+            nsamples += 1
+
+        xdata = calcFrequencies(nsamples, overlay.tr, np.float32)

         return xdata, ydata
```

`<your-python-env>` will differ depending on how you have installed FSLeyes. If you are using FSLeyes from wtihin a FSL >= 6.0.6 installation, `<your-python-env>` will be `$FSLDIR`. For older versions of FSL, it will be `$FSLDIR/fslpython/`. The Python version `<X.Y>` will also depend on your specific installation - for FSL >= 6.0.6, it will be `3.10`.

**Comment by Nagle2036** [comment]
Hi @pauldmccarthy, thanks for providing this workaround. I've made the changes in powerspectrumseries.py and it all seems to be working! Thanks so much for helping me fix this issue so quickly, much appreciated.

Alex

**Comment by pauldmccarthy** [comment]
Fixed in FSLeyes 1.7.0
